### PR TITLE
[mlir][inliner] optimize self-recursive function detection

### PR DESCRIPTION
**Pull request by HerrCai0907**
It helps the iteration finish as quickly as possible.
The original implement depends on the target of edge in `targetCall`, which is maintained by call graph and will not update during inlining. This PR wants to prepare a recursiveCallRegions before each iteration, then It can dynamically update.

Here is an example:
```mlir
func.func @b0() {
  func.call @b1() : () -> ()
  return
}
func.func @b1() {
  func.call @b0() : () -> ()
  return
}
```

The original inline debug log looks like:
```
* Inliner: Initial calls in SCC are: {
  0. func.call @b0() : () -> (),
  1. func.call @b1() : () -> (),
}
* Inlining call: 0. func.call @b0() : () -> ()
* new inlineHistory entry: 0. [func.call @b0() : () -> (), root]
* new call 2 {func.call @b0() : () -> ()}
   with historyID = 0, added due to inlining of
  call {func.call @b0() : () -> ()}
 with historyID = root
* Inlining call: 1. func.call @b1() : () -> ()
* new inlineHistory entry: 1. [func.call @b1() : () -> (), root]
* new call 3 {func.call @b1() : () -> ()}
   with historyID = 1, added due to inlining of
  call {func.call @b1() : () -> ()}
 with historyID = root
* Not inlining call: 2. func.call @b1() : () -> ()
* Not inlining call: 3. func.call @b1() : () -> ()
* Inliner: Initial calls in SCC are: {
  0. func.call @b1() : () -> (),
  1. func.call @b1() : () -> (),
}
* Not inlining call: 0. func.call @b1() : () -> ()
* Inlining call: 1. func.call @b1() : () -> ()
* new inlineHistory entry: 0. [func.call @b1() : () -> (), root]
* new call 2 {func.call @b1() : () -> ()}
   with historyID = 0, added due to inlining of
  call {func.call @b1() : () -> ()}
 with historyID = root
* Not inlining call: 2. func.call @b1() : () -> ()
* Inliner: Initial calls in SCC are: {
  0. func.call @b1() : () -> (),
  1. func.call @b1() : () -> (),
}
* Not inlining call: 0. func.call @b1() : () -> ()
* Inlining call: 1. func.call @b1() : () -> ()
* new inlineHistory entry: 0. [func.call @b1() : () -> (), root]
* new call 2 {func.call @b1() : () -> ()}
   with historyID = 0, added due to inlining of
  call {func.call @b1() : () -> ()}
 with historyID = root
* Not inlining call: 2. func.call @b1() : () -> ()
* Inliner: Initial calls in SCC are: {
  0. func.call @b1() : () -> (),
  1. func.call @b1() : () -> (),
}
* Not inlining call: 0. func.call @b1() : () -> ()
* Inlining call: 1. func.call @b1() : () -> ()
* new inlineHistory entry: 0. [func.call @b1() : () -> (), root]
* new call 2 {func.call @b1() : () -> ()}
   with historyID = 0, added due to inlining of
  call {func.call @b1() : () -> ()}
 with historyID = root
* Not inlining call: 2. func.call @b1() : () -> ()
* Inliner: Initial calls in SCC are: {
}
```

After optimization:
```
* Inliner: Initial calls in SCC are: {
  0. func.call @b0() : () -> (),
  1. func.call @b1() : () -> (),
}
* Inlining call: 0. func.call @b0() : () -> ()
* new inlineHistory entry: 0. [func.call @b0() : () -> (), root]
* new call 2 {func.call @b0() : () -> ()}
   with historyID = 0, added due to inlining of
  call {func.call @b0() : () -> ()}
 with historyID = root
* Inlining call: 1. func.call @b1() : () -> ()
* new inlineHistory entry: 1. [func.call @b1() : () -> (), root]
* new call 3 {func.call @b1() : () -> ()}
   with historyID = 1, added due to inlining of
  call {func.call @b1() : () -> ()}
 with historyID = root
* Not inlining call: 2. func.call @b1() : () -> ()
* Not inlining call: 3. func.call @b1() : () -> ()
* Inliner: Initial calls in SCC are: {
  0. func.call @b1() : () -> (),
  1. func.call @b1() : () -> (),
}
* Not inlining call: 0. func.call @b1() : () -> ()
* Not inlining call: 1. func.call @b1() : () -> ()
* Inliner: Initial calls in SCC are: {
}
```

### CODE REVIEW

File: mlir/lib/Transforms/Utils/Inliner.cpp [hunk 1]
```diff
@@ -621,6 +623,12 @@ Inliner::Impl::inlineCallsInSCC(InlinerInterfaceImpl &inlinerIface,
     llvm::dbgs() << "}\n";
   });
 
+  llvm::SmallPtrSet<Region *, 16U> recursiveCallRegions{};
```
[the remark this answers is not in the document]
Comment (HerrCai0907) [reply mlir/lib/Transforms/Utils/Inliner.cpp:626]: It doesn't have default size value for `SmallPtrSet`. I just pick 16 between 0 to 32.

**Comment by HerrCai0907** [comment]
friendly ping @joker-eph

**Comment by HerrCai0907** [comment]
friendly ping @joker-eph

File: mlir/lib/Transforms/Utils/Inliner.cpp [hunk 2]
```diff
@@ -705,24 +715,30 @@ Inliner::Impl::inlineCallsInSCC(InlinerInterfaceImpl &inlinerIface,
...
-                     }) > 0)
-    return false;
-
-  // Don't allow inlining if the target is an ancestor of the call. This
-  // prevents inlining recursively.
   Region *callableRegion = resolvedCall.targetNode->getCallableRegion();
-  if (callableRegion->isAncestor(resolvedCall.call->getParentRegion()))
+
+  // Don't allow inlining this following cases to prevent inlining recursively.
+  // 1. target has at least an edge back to itself in original call graph.
```
Comment (joker-eph) [inline mlir/lib/Transforms/Utils/Inliner.cpp:736]: This seems overly conservative I would think, wouldn't it prevent to inline a case like:

```
void foo(int value) {
  if (value > 20) foo(value-1);
}
void bar() {
  foo(10);
}
```

Here foo is self-recursive, but inlining it into bar is still desirable.

File: mlir/lib/Transforms/Utils/Inliner.cpp [hunk 3]
```diff
@@ -705,24 +715,30 @@ Inliner::Impl::inlineCallsInSCC(InlinerInterfaceImpl &inlinerIface,
...
+  // 2. target has call instructions call itself after pervious inlining.
```
Comment (joker-eph) [inline mlir/lib/Transforms/Utils/Inliner.cpp:737]:
```suggestion
  // 2. target has call instructions call itself after previous inlining.
```

File: mlir/lib/Transforms/Utils/Inliner.cpp (hunk 2 above)
Comment (HerrCai0907) [reply mlir/lib/Transforms/Utils/Inliner.cpp:736]: Then how can we avoid infinite loop, `foo` can be inlined again and again.
I meet a issue, when I use `inliner` in some code like following example, exponential explosion happen.
```
void foo(int value) {
foo(value-1);
foo(value-1);
}

void bar() {
foo(10);
foo(10);
}
```

**Pull request closed without merging**